tfhe/src/shortint/casting_key/mod.rs

```diff
@@ -0,0 +1,159 @@
+    /// depending on input.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use tfhe::shortint::parameters::{PARAM_MESSAGE_1_CARRY_1, PARAM_MESSAGE_2_CARRY_2};
+    /// use tfhe::shortint::{gen_keys, CastingKey};
+    ///
+    /// // Generate the client keys and server keys:
+    /// let (ck1, sk1) = gen_keys(PARAM_MESSAGE_1_CARRY_1);
+    /// let (ck2, sk2) = gen_keys(PARAM_MESSAGE_2_CARRY_2);
+    ///
+    /// // Generate the server key:
+    /// let ksk = CastingKey::new((&ck1, &sk1), (&ck2, &sk2));
+    /// ```
+    pub fn new(key_pair_1: (&ClientKey, &ServerKey), key_pair_2: (&ClientKey, &ServerKey)) -> Self {
+        // Creation of the key switching key
+        let key_switching_key = ShortintEngine::with_thread_local_mut(|engine| {
+            engine.new_key_switching_key(key_pair_1.0, key_pair_2.0)
+        });
+
+        let full_message_modulus_1 =
+            key_pair_1.0.parameters.carry_modulus().0 * key_pair_1.0.parameters.message_modulus().0;
+        let full_message_modulus_2 =
+            key_pair_2.0.parameters.carry_modulus().0 * key_pair_2.0.parameters.message_modulus().0;
+        if (full_message_modulus_1 & (full_message_modulus_1 - 1)) != 0
+            || (full_message_modulus_2 & (full_message_modulus_2 - 1)) != 0
+        {
+            panic!("Cannot create casting key if the full messages moduli are not a power of 2");
+        }
+
+        let mut nb_bits_1 = 1i8;
+        while full_message_modulus_1 != 1 << nb_bits_1 {
+            nb_bits_1 += 1;
+        }
+        let mut nb_bits_2 = 1i8;
+        while full_message_modulus_2 != 1 << nb_bits_2 {
+            nb_bits_2 += 1;
+        }
+
+        // Pack the keys in the casting key set:
+        Self {
+            key_switching_key: key_switching_key.unwrap(),
+            dest_server_key: key_pair_2.1.clone(),
+            src_server_key: key_pair_1.1.clone(),
+            cast_rshift: nb_bits_2 - nb_bits_1,
+        }
+    }
+
+    /// Cast a ciphertext from the source parameter set to the dest parameter set,
+    /// using provided &mut.
+    ///
+    /// # Example (the following code won't actually run because this function is private)
+    ///
+    /// ```rust
+    /// use tfhe::shortint::parameters::{PARAM_MESSAGE_1_CARRY_1, PARAM_MESSAGE_2_CARRY_2};
+    /// use tfhe::shortint::{gen_keys, CastingKey};
+    ///
+    /// // Generate the client keys and server keys:
+    /// let (ck1, sk1) = gen_keys(PARAM_MESSAGE_1_CARRY_1);
+    /// let (ck2, sk2) = gen_keys(PARAM_MESSAGE_2_CARRY_2);
+    ///
+    /// // Generate the server key:
+    /// let ksk = CastingKey::new((&ck1, &sk1), (&ck2, &sk2));
+    ///
+    /// let cipher = ck1.encrypt(1);
+    /// let cipher_2 = sk2.create_trivial(0);
+    /// ksk.cast_assign(&cipher, &mut cipher_2);
+    /// ```
+    pub fn cast_assign<OpOrder: PBSOrderMarker>(
+        &self,
+        ct: &CiphertextBase<OpOrder>,
+        ct_dest: &mut CiphertextBase<OpOrder>,
+    ) {
```

Comment:
Same as in boolean, generaly we take the dest as first argument in our assign functions